### PR TITLE
Support for Sending/Receivng Ethernet Packets. 

### PR DESCRIPTION
**Pull request by umarbinzahid**
This PR includes the following changes:
1. Adds support for sending and receiving Ethernet packets. This support is feature-gated and can be enabled by passing the `ethernet` flag.
2. Adds support for booting the device in `Ethernet` mode.
3. Adds a `BootMode` enum to select between `Normal` and `Ethernet` modes.
4. Adds unit tests for the newly introduced APIs.

## Summary by Sourcery

Add feature-gated Ethernet mode support, including boot-time selection and APIs for sending and receiving raw Ethernet packets, plus associated event handling and tests.

New Features:
- Introduce a BootMode enum to select between normal and Ethernet boot modes for the WiFi module.
- Add public APIs on WincClient to start the module in Ethernet mode and to send and receive raw Ethernet frames when the ethernet feature is enabled.
- Expose Ethernet RX metadata via a new EthernetRxInfo type and corresponding callback in the event listener pipeline.

Enhancements:
- Refine boot configuration to derive chip configuration from BootMode combined with a shared default config constant.
- Handle previously unhandled WiFi HIF responses by logging and returning explicit InvalidHifResponse errors instead of panicking.

Build:
- Add an optional ethernet Cargo feature flag to gate Ethernet-mode functionality.

Tests:
- Add unit tests for Ethernet RX reply parsing, Ethernet packet send/receive client APIs, and timeout handling in Ethernet receive logic.

## Summary by CodeRabbit

* **New Features**
  * Optional ethernet support (feature-flagged): send/receive raw Ethernet frames and an Ethernet boot mode.
  * Boot-mode-driven startup: explicit Normal and Ethernet start paths selectable at startup.

* **Refactor**
  * Wi‑Fi startup reorganized into a two-step implementation with public delegators for Normal and Ethernet modes.

* **Tests**
  * Unit tests for ethernet send/receive behavior, buffering, timeouts, and edge cases.

* **Documentation**
  * Updated docs to describe boot-mode-driven initialization.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: winc-rs/src/stack/socket_callbacks.rs [hunk 1]
```diff
@@ -18,6 +18,9 @@ use crate::manager::{SslCallbackInfo, SslResponse};
 #[cfg(feature = "experimental-ecc")]
 use crate::manager::EccRequest;
 
+#[cfg(feature = "ethernet")]
+use crate::manager::EthernetRxInfo;
+
 use super::sock_holder::{SockHolder, SocketStore};
 use crate::{debug, error, info};
 use crate::{socket::Socket, Ipv4AddrFormatWrapper};
```
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/stack/socket_callbacks.rs]: This is fine because until all data is read from the module, the `RX` flag is not written in `Regs::WifiHostRcvCtrl0`. As a result, a new `Ethernet RX info` packet will not be sent from the module.

File: winc-rs/src/manager.rs [hunk 2]
```diff
@@ -1741,6 +1748,69 @@ impl<X: Xfer> Manager<X> {
...
+        self.chip
+            .dma_block_write(self.not_a_reg_ctrl_4_dma + HIF_HEADER_OFFSET as u32, &req)?;
+
+        self.chip.dma_block_write(
+            self.not_a_reg_ctrl_4_dma + ETHERNET_HEADER_OFFSET as u32,
+            net_pkt,
+        )?;
+
+        self.write_ctrl3(self.not_a_reg_ctrl_4_dma)
+    }
```
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/manager.rs]: Check is added to validate the buffer length.

File: winc-rs/src/client/ethernet.rs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+// limitations under the License.
+
+use super::{StackError, WincClient, Xfer};
+use embedded_nal::nb;
+
+/// 1 second timeout to read etherent packet.
+const ETHERNET_RX_TIMEOUT_MSEC: u32 = 1000;
+
+impl<X: Xfer> WincClient<'_, X> {
+    /// Tries to read an Ethernet packet fromt he module within a specified timeout.
```
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/client/ethernet.rs]: Typo is corrected.

File: winc-rs/src/client/ethernet.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        timeout: Option<u32>,
+    ) -> nb::Result<usize, StackError> {
+        match &mut self.callbacks.eth_rx_info {
+            None => {
+                self.callbacks.eth_rx_info = Some(None);
+                if let Some(timeout) = timeout {
+                    self.operation_countdown = timeout;
+                } else {
+                    self.operation_countdown = ETHERNET_RX_TIMEOUT_MSEC;
+                }
```
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/client/ethernet.rs]: This has been updated per the suggested code; however, this change shouldn’t be required because `self.poll_loop_delay_us` should already have the correct value to generate the appropriate polling loop timeout. Since changing it would also affect other operations, it will be addressed in a separate PR.

File: winc-rs/src/manager.rs [hunk 5]
```diff
@@ -1741,6 +1748,69 @@ impl<X: Xfer> Manager<X> {
...
+        address: u32,
+        data: &mut [u8],
+        rx_done: bool,
+    ) -> Result<(), Error> {
+        self.chip.dma_block_read(address, data)?;
+        // clear RX if no more data is available to read.
+        if rx_done {
+            let reg = self.chip.single_reg_read(Regs::WifiHostRcvCtrl0.into())?;
+            self.chip
+                .single_reg_write(Regs::WifiHostRcvCtrl0.into(), reg | 2)?;
```
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/manager.rs]: This will be addressed in a separate PR.

File: winc-rs/src/stack/socket_callbacks.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/stack/socket_callbacks.rs]: The condition to store data only when `eth_rx_info` is `Some(None)` has been removed to avoid the possibility of missing any packet, which is unlikely because the first packet is expected to be received when the user/TCP stack sends any Ethernet packet.

Also, once `eth_rx_info` is received, no new packet is expected to arrive until the RX_DONE flag is written to `Regs::WifiHostRcvCtrl0`.

**Review by kaidokert — APPROVED** [review]
Overally looks all good - lets get it rebased and ready to merge

File: winc-rs/src/client/ethernet.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+                        self.callbacks.eth_rx_info = None;
+                        return Err(nb::Error::Other(StackError::GeneralTimeout));
+                    }
+                    self.operation_countdown -= 1;
+                }
+            }
+        }
+        self.dispatch_events()?;
+        Err(nb::Error::WouldBlock)
+    }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply winc-rs/src/client/ethernet.rs]: I agree, theres SPI layer crc checking already and ethernet frame level crcs as well. No risk here

File: winc-rs/src/async_client/module.rs [hunk 7]
```diff
@@ -15,7 +16,7 @@ impl<X: Xfer> AsyncClient<'_, X> {
         self.callbacks.borrow_mut().state = WifiModuleState::Starting;
         self.manager.borrow_mut().set_crc_state(true);
 
-        let mut state = Default::default();
+        let mut state = BootState::new(BootMode::Normal);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply winc-rs/src/async_client/module.rs]: Async client is pretty incomplete anyway. Meta issue is #25 but i think we'll need a more specific ones.

**Review by kaidokert — APPROVED** [review]
Seems solid ! Great addition
